### PR TITLE
[master < T597] Add restoring of replication roles upon database startup

### PR DESCRIPTION
**Pull request by Josipmrden**
[master < Task] PR
- [X] Check, and update documentation if necessary
- [ ] Provide the full content or a guide for the final git message

To keep docs changelog up to date, one more thing to do:
- [X] Write a release note here
- [ ] Tag someone from docs team in the comments

Release note: Replication role is now restored upon the startup of database. Replicas are no longer restored as main instances with manual effort needed to configure the replica again.

### CODE REVIEW

**Comment by Josipmrden** [comment]
@vpavicic tagging for release note

**Comment by vpavicic** [comment]
Can we just say this:

Replica instances no longer restart as main instances, but restore their original role.

**Comment by Josipmrden** [comment]
yes @vpavicic

**Comment by Josipmrden** [comment]
@gitbuda  and @antoniofilipovic  I'd need reviews

File: src/storage/v2/replication/replication_persistence_helper.hpp [hunk 1]
```diff
@@ -23,18 +23,23 @@
 
 namespace memgraph::storage::replication {
 
-struct ReplicaStatus {
+static const std::string kReservedReplicationRoleName{"__replication_role"};
```
Comment (antoniofilipovic) [inline src/storage/v2/replication/replication_persistence_helper.hpp:26]: Is there some reason not to use inline constexpr?

File: src/storage/v2/replication/replication_persistence_helper.cpp [hunk 2]
```diff
@@ -70,6 +75,13 @@ std::optional<ReplicaStatus> JSONToReplicaStatus(nlohmann::json &&data) {
       data.at(kSSLKeyFile).get_to(replica_status.ssl->key_file);
       data.at(kSSLCertFile).get_to(replica_status.ssl->cert_file);
     }
+
+    if (data.find(kReplicationRole) != data.end()) {
+      auto role = replication::ReplicationRole::MAIN;  // initializing it to dummy value because of clang-tidy
+      data.at(kReplicationRole).get_to(role);          // here happens the assigning of the role
```
Comment (antoniofilipovic) [inline src/storage/v2/replication/replication_persistence_helper.cpp:81]: replica_status.role = replication::ReplicationRole::MAIN;
data.at(kReplicationRole).get_to(replica_status.role.value());   

https://en.cppreference.com/w/cpp/utility/optional/value

I think this should work. `value()` returns reference if there is value inside, and get_to will store value on that reference, so all good, right?

File: src/storage/v2/storage.cpp [hunk 3]
```diff
@@ -323,13 +323,6 @@ Storage::Storage(Config config)
       uuid_(utils::GenerateUUID()),
       epoch_id_(utils::GenerateUUID()),
       global_locker_(file_retainer_.AddLocker()) {
```
Comment (antoniofilipovic) [inline src/storage/v2/storage.cpp:325]: Why did you delete this `warn` comment?

File: src/storage/v2/storage.cpp [hunk 4]
```diff
@@ -418,14 +411,29 @@ Storage::Storage(Config config)
...
-  if (config_.durability.restore_replicas_on_startup) {
-    spdlog::info("Replica's configuration will be stored and will be automatically restored in case of a crash.");
+  if (config_.durability.restore_replication_state_on_startup) {
+    spdlog::info("Replication configuration will be stored and will be automatically restored in case of a crash.");
     utils::EnsureDirOrDie(config_.durability.storage_directory / durability::kReplicationDirectory);
     storage_ =
         std::make_unique<kvstore::KVStore>(config_.durability.storage_directory / durability::kReplicationDirectory);
-    RestoreReplicas();
+
+    RestoreReplicationRole();
```
Comment (antoniofilipovic) [inline src/storage/v2/storage.cpp:420]: I see here we are restoring replication role on startup, but why are we restoring replica only for MAIN?

File: src/storage/v2/replication/replication_persistence_helper.hpp (hunk 1 above)
Comment (Josipmrden) [reply src/storage/v2/replication/replication_persistence_helper.hpp:26]: changed to string_view and inline

File: src/storage/v2/storage.cpp (hunk 3 above)
Comment (Josipmrden) [reply src/storage/v2/storage.cpp:337]: it is actually further down

Comment (Josipmrden) [reply src/storage/v2/storage.cpp:337]: after the restoring of the replica role

File: src/storage/v2/storage.cpp (hunk 4 above)
Comment (Josipmrden) [reply src/storage/v2/storage.cpp:439]: we are restoring replication role, and restoring corresponding replicas if it's the main

**Review by antoniofilipovic — APPROVED** [review]
Looks good to me

**Comment by Josipmrden** [comment]
@vpavicic
